merkle/smt/hstar3.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package smt contains the implementation of the sparse Merkle tree logic.
+package smt
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/google/trillian/storage/tree"
+)
+
+// NodeStorage reads and writes sparse Merkle tree node hashes.
+type NodeStorage interface {
+	// Get returns the hash of the given node.
+	Get(id tree.NodeID2) ([]byte, error)
+	// Set sets the hash of the given node.
+	Set(id tree.NodeID2, hash []byte)
+}
+
+// HashChildrenFn computes a node hash based on its child nodes' hashes.
+type HashChildrenFn func(left, right []byte) []byte
+
+// NodeUpdate represents an update of a node hash in HStar3 algorithm.
+type NodeUpdate struct {
+	ID   tree.NodeID2
+	Hash []byte
+}
+
+// HStar3 is a faster non-recursive HStar2.
+//
+// TODO(pavelkalinnikov): Swap in the code, and document it properly.
+type HStar3 struct {
+	hash  HashChildrenFn
+	depth uint
+}
+
+// NewHStar3 returns a new instance of HStar3.
+func NewHStar3(hash HashChildrenFn, depth uint) HStar3 {
+	return HStar3{hash: hash, depth: depth}
+}
+
+// Prepare sorts the updates slice for it to be usable by HStar3. It also
+// verifies that the nodes are placed at the required depth, and there are no
+// ID duplicates amongh them.
+func (h HStar3) Prepare(updates []NodeUpdate) error {
+	for i := range updates {
+		if d, want := updates[i].ID.BitLen(), h.depth; d != want {
+			return fmt.Errorf("upd #%d: invalid depth %d, want %d", i, d, want)
+		}
+	}
+	sort.Slice(updates, func(i, j int) bool {
+		return compareHorizontal(updates[i].ID, updates[j].ID)
+	})
+	for i, last := 0, len(updates)-1; i < last; i++ {
+		if id := updates[i].ID; id == updates[i+1].ID {
+			return fmt.Errorf("duplicate ID: %v", id)
+		}
+	}
+	return nil
+}
+
+// GetReadSet returns the set of all the node IDs that HStar3 requires to read
+// in order to apply the passed-in updates and propagate node hash changes up
+// to the root(top) node(s) at the specified depth. The updates must have been
+// processed by Prepare.
+//
+// Note: The returned map could have bool value type, but []byte allows the
+// caller to reuse this map for filling in the hashes for the Update method.
+//
+// TODO(pavelkalinnikov): Return only tile IDs.
+func (h HStar3) GetReadSet(updates []NodeUpdate, top uint) map[tree.NodeID2][]byte {
```

Comment:
This name is properly confusing. It's a bit like Reading Festival - is it a festival with lots of books, or is it a festival in Reading? Or both?

I initially read this as 3 verbs: "Get x, Read x, Set x", but it's intended to be parsed as "Get the {nodes} to read".

This API surface area is a bit large for what is required here. I would propose an inversion of control - have a single method which takes objects that allow these nodes to be fetched in batches. That also avoids the strange artifact where this function is creating an overly ambitious return type because it can predict what its clients are going to do next.

Rough idea of how my thoughts could be implemented
```
NodeStorageBuilder {
  nodes map[tree.NodeID2][]byte
}
func (nsb NodeStorageBuilder) Prefetch(n tree.NodeID2) {
  nodes[n] = nil
}
func (nsb NodeStorageBuilder) BuildStorage() NodeStorage {
  return storageDependantFetchMethod(nsb.nodes)
}


UberUpdate(updates []NodeUpdate, top uint, nsb NodeStorageBuilder) {
  prepare(updates)
  getReadSet(nsb)
  update(updates, top, nsb.BuildStorage())
}
```
In the case that you don't want to prefetch, the NSB would do nothing for Prefetch, and then return something with a direct line to the DB for BuildStorage. For the (sane) case where you want to prefetch things, the only place that would talk to the DB is what I've called storageDependantFetchMethod above, and the NodeStorage returned wouldn't have any connection to the DB which prevents prefetch misses from silently succeeding.

Reply:
`Prepare` or similar?

Reply:
`Prepare` sounds good. Done.